python/lsst/ts/m2com/mock/mock_control_closed_loop.py

```diff
@@ -0,0 +1,860 @@
+
+    def reset_force_offsets(self):
+        """Reset the force offsets.
+
+        This will put the applied force to be zero.
+        """
+
+        self.axial_forces["applied"] = np.zeros(NUM_ACTUATOR - NUM_TANGENT_LINK)
+        self.tangent_forces["applied"] = np.zeros(NUM_TANGENT_LINK)
+
+    def get_net_forces_total(self):
+        """Get the total net forces in Newton.
+
+        Returns
+        -------
+        `dict`
+            Total net forces in Newton.
+        """
+        return self._calc_net_forces_total(
+            self.axial_forces["measured"], self.tangent_forces["measured"]
+        )
+
+    def _calc_net_forces_total(self, axial_forces, tangent_forces):
+        """Calculate the total net forces in Newton.
+
+        Parameters
+        ----------
+        axial_forces : `numpy.ndarray`
+            Axial actuator forces in Newton.
+        tangent_forces : `numpy.ndarray`
+            Tangent actuator forces in Newton.
+
+        Returns
+        -------
+        `dict`
+            Total net forces in Newton.
+        """
+
+        angle = np.deg2rad(self._cell_geom["locAct_tangent"])
+
+        fx = np.sum(np.cos(angle) * tangent_forces)
+        fy = np.sum(np.sin(angle) * tangent_forces)
+        fz = np.sum(axial_forces)
+
+        return {"fx": fx, "fy": fy, "fz": fz}
+
+    def get_net_moments_total(self):
+        """Get the total net moments in Newton * meter.
+
+        Returns
+        -------
+        `dict`
+            Total net moments in Newton * meter.
+        """
+        return self._calc_net_moments_total(
+            self.axial_forces["measured"], self.tangent_forces["measured"]
+        )
+
+    def _calc_net_moments_total(self, axial_forces, tangent_forces):
+        """Calculate the total net moments in Newton * meter.
+
+        Parameters
+        ----------
+        axial_forces : `numpy.ndarray`
+            Axial actuator forces in Newton.
+        tangent_forces : `numpy.ndarray`
+            Tangent actuator forces in Newton.
+
+        Returns
+        -------
+        `dict`
+            Total net moments in Newton * meter.
+        """
+
+        location_axial_actuator = np.array(self._cell_geom["locAct_axial"])
+        mx = np.sum(axial_forces * location_axial_actuator[:, 1])
+        my = np.sum(axial_forces * location_axial_actuator[:, 0])
+        mz = np.sum(tangent_forces) * self._cell_geom["radiusActTangent"]
+
+        return {"mx": mx, "my": my, "mz": mz}
+
+    def get_force_balance(self):
+        """Get the data of force balance system. This contains the net forces
+        (in Newton) and net moments (in Newton * meter).
+
+        Returns
+        -------
+        force_balance : `dict`
+            Data of the force balance system.
+        """
+
+        axial_forces = self.axial_forces["hardpointCorrection"]
+        tangent_forces = self.tangent_forces["hardpointCorrection"]
+
+        net_forces = self._calc_net_forces_total(axial_forces, tangent_forces)
```

Comment:
I am not sure how widely you are using `self._calc_net_forces_total` and `self._calc_net_momments_total`, so take this comment with a grain of salt. 

But, consider what is happening when you call them like this. You are creating a variable with the data from a class member and then passing all that data to a function of the same class.. There's a lot of data allocation and transport going on here. 

Now, if you are only using it to compute things from class data that are in `self.axial_forces` and `self.tangent_forces`, you could potentially change the method to receive the dictionary key and use the data from the class directly.. For example:

```
def _calc_net_forces_total(self, force):
        angle = np.deg2rad(self._cell_geom["locAct_tangent"])

         fx = np.sum(np.cos(angle) * self.tangent_forces[force])
         fy = np.sum(np.sin(angle) * self.tangent_forces[force])
         fz = np.sum(self.axial_forces[force])

         return {"fx": fx, "fy": fy, "fz": fz}
```

This way there's no data allocation and transport happening, just local data access... 

But again, this only works if you are only using the method with the internal data.